pkg/runtime/sequencer/cache.go

```diff
@@ -2,15 +2,198 @@ package sequencer
+	Key          string
+	Size         int
+	InUseBuffer  *Buffer
+	BackUpBuffer *Buffer
+	lock         sync.Mutex
+	Store        sequencer.Store
+}
+
+type Buffer struct {
+	from int64
+	to   int64
+}
+
+func NewDoubleBuffer(key string, store sequencer.Store) *DoubleBuffer {
+
+	d := &DoubleBuffer{
+		Key:   key,
+		Size:  defaultSize,
+		Store: store,
+	}
+
+	return d
+}
+
+//init double buffer
+func (d *DoubleBuffer) init() error {
+
+	buffer, err := d.getNewBuffer()
+	if err != nil {
+		return err
+	}
+
+	d.InUseBuffer = buffer
+
+	return nil
+}
+
+//getId next id
+func (d *DoubleBuffer) getId() (int64, error) {
+
+	if d.InUseBuffer == nil {
+		return 0, errors.New("[DoubleBuffer] Get error: InUseBuffer nil ")
+	}
+
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	//check swap
+	if d.InUseBuffer.from > d.InUseBuffer.to {
+		err := d.swap()
+		if err != nil {
+			return 0, err
+		}
+	}
+	next := d.InUseBuffer.from
+	d.InUseBuffer.from++
+
+	//when InUseBuffer id more than half used, initialize BackUpBuffer.
+	//equal make sure only one thread enter
+	if d.InUseBuffer.to-d.InUseBuffer.from == defaultLimit {
+		utils.GoWithRecover(func() {
+
+			//lock to avoid data race
+			d.lock.Lock()
+			defer d.lock.Unlock()
+			//one case: here and swap are performed simultaneously,
+			//swap fast,no check will cover old BackUpBuffer
+			if d.BackUpBuffer != nil {
+				return
+			}
+			buffer, err := d.getNewBuffer()
+			if err != nil {
+				log.DefaultLogger.Errorf("[DoubleBuffer] [getNewBuffer] error: %v", err)
+				return
+			}
+
+			d.BackUpBuffer = buffer
+
+		}, nil)
+	}
+
+	return next, nil
+}
+
+//swap InUseBuffer and BackUpBuffer, must be locked
+func (d *DoubleBuffer) swap() error {
+
+	//check again
+	if d.BackUpBuffer == nil {
+		buffer, err := d.getNewBuffer()
+		if err != nil {
+			return err
+		}
+		d.BackUpBuffer = buffer
+	}
+
+	d.InUseBuffer = d.BackUpBuffer
+	d.BackUpBuffer = nil
+	return nil
+}
+
+//getNewBuffer return a new segment
+func (d *DoubleBuffer) getNewBuffer() (*Buffer, error) {
+	support, result, err := d.Store.GetSegment(&sequencer.GetSegmentRequest{
+		Key:  d.Key,
+		Size: d.Size,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if !support {
+		return nil, errors.New("[DoubleBuffer] unSupport Segment id")
+	}
+	return &Buffer{
+		from: result.From,
+		to:   result.To,
+	}, nil
+}
+
+var BufferCatch = map[string]*DoubleBuffer{}
+
+//common lock is enough ？
+var rwLock sync.RWMutex
+
 func GetNextIdFromCache(ctx context.Context, store sequencer.Store, req *sequencer.GetNextIdRequest) (bool, int64, error) {
-	// TODO
-	//	1. check cache
-	//	2. load cache with lock
-	//	2.1. return if not support
-	//	2.2. add segment into cache
-	//	2.3. return result
-	return false, 0, nil
+
+	//check support
+	support, _, _ := store.GetSegment(&sequencer.GetSegmentRequest{
+		Key:  req.Key,
+		Size: 0,
+	})
+
+	if !support {
+		return false, 0, nil
+	}
+
+	var d *DoubleBuffer
+	var err error
+	if _, ok := BufferCatch[req.Key]; !ok {
```

Comment:
I find a new problem here: BufferCatch's key is just the requested key without component name,but unfortunately different components might have same key, so they might affect each other   :3

Reply:
@seeflood  can one layotto used by  more than one application ?

Reply:
@ZLBer Currently no, but maybe yes in the future.
But even if layotto serve only one application,this app can use two sequencers with two different `storeName`,and it will lead to conflicts here

Reply:
@seeflood  see this question again,  an easy way to solve this is add componentName before key in  `runtime_sequencer.GetModifiedKey` method.

Reply:
@ZLBer You are right !
We can modify keys there like`GetModifiedLockKey` in lock_config.go or `GetModifiedStateKey` in compatibility.go
Would u like to do it? Or let me submit a new PR to fix it after this PR has been merged

Reply:
@seeflood  i will fix this

Reply:
@ZLBer Cool
You can make it support different naming strategy ,like `GetModifiedLockKey` in lock_config.go